File: 46.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        if(nums.length == 1) {
            ArrayList<List<Integer>> a = new ArrayList<List<Integer>>();
            ArrayList<Integer> aa = new ArrayList<Integer>();
            aa.add(nums[0]);
            a.add(aa);
            return a;
        }
        ArrayList<List<Integer>> list = new ArrayList<List<Integer>>();
        for(int i = 0; i < nums.length; ++i) {
            // get permutations of every element but nums[i]
            int[] arr = new int[nums.length-1];
            int c = 0;
            for(int n : nums) {
                if(n != nums[i])
                    arr[c++] = n;
            }
            List<List<Integer>> l = permute(arr);
            for(List<Integer> ll : l) {
                ArrayList<Integer> a = new ArrayList<Integer>();
                a.add(nums[i]);
                for(int n : ll) {
                    a.add(n);
                }
                list.add(a);
            }
        }
        return list;
    }
}

File: 46.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        if(nums.size() == 1) {
            vector<vector<int>> v = {{nums[0]}};
            return v;
        }
        vector<vector<int>> a;
        for(int i = 0; i < nums.size(); ++i) {
            vector<int> b;
            for(int j = 0; j < nums.size(); ++j) {
                if(j == i)
                    continue;
                b.push_back(nums[j]);
            }
            vector<vector<int>> c = permute(b);
            for(int j = 0; j < c.size(); ++j) {
                vector<int> p = c[j];
                vector<int> d = {nums[i]};
                for(int k = 0; k < p.size(); ++k)
                    d.push_back(p[k]);
                a.push_back(d);
            }
        }
        return a;
    }
};
